src/expr/type_node.h

```diff
@@ -530,6 +540,33 @@ class CVC5_EXPORT TypeNode
    */
   bool isFunctionLike() const;
 
+  /**
+   * Is instance of, returns true if this type is equivalent to the join
+   * (see TypeNode::join) of itself and t.
+   */
+  bool isInstanceOf(const TypeNode& t) const;
+  /**
+   * Is comparable to type t, returns true if this type and t have a non-null
+   * join (see TypeNode::join).
+   */
+  bool isComparableTo(const TypeNode& t) const;
+  /**
+   * Join with type. This returns the most specific type that is an instance
+   * of both this and t, or null if this type and t are incompatible.
+   *
+   * For example:
+   * ?BitVec join ? = ?BitVec
+   * (Array ?BitVec Int) join (Array (_ BitVec 4) ?) = (Array (_ BitVec 4) Int)
+   * (Array ? Int) join (Array ? Real) = null.
+   */
+  TypeNode join(const TypeNode& t) const;
```

Comment:
How is the partial order defined here?
The way I see it is that `?BitVec <= ?`, because in sets, the set `?BitVec` is a subset of `?`. 
Therefore I think the name of this function should be `meet` since the `join` is the [least upper bound](https://mathworld.wolfram.com/Join.html). 
To avoid confusion which one is which, how about just using the short names `supremum, infimum` or the corresponding long names `leastUpperBound, greatest lower bound` respectively?

Reply:
Sounds good, I will use `leastUpperBound` instead of `join` / `greatestLowerBound` instead of `meet`.

Reply:
The partial order is defined in terms of matching, a type T1 <= T2 if T2 matches T1, where gradual types in T1 match (possibly) more specific ones in T2.